### ISSUE TITLE
SAML message intended destination endpoint ERROR

### ISSUE THREAD

**Issue by greven** [pac4j/spark-pac4j#4]
The Spark-Pac4J library regarding SAML doesn't' seem to register correctly the callback URL when receiving the SAML Response.

The error I get is: 

    ERROR org.opensaml.common.binding.decoding.BaseSAMLMessageDecoder - SAML message intended destination endpoint 'https://localhost/callback?client_name=Saml2Client' did not match the recipient endpoint 'https://localhost/callback'

As described here: [Pac4J-GoogleGroups](https://groups.google.com/d/msg/pac4j-users/pBlWwnYqoJk/kegzWKXbgsUJ).

**Comment by leleuj** [comment]
Hi,

The /callback endpoint is shared among clients, so to know the client used,
a parameter (client_name) is added. So the real callback url in your
application from your SAML IdP is /callback?client_name=Saml2Client
(Saml2Client being the name of your SAML client, be careful, it might be
SAML2Client with pac4j v1.7.1).

So I guess your SAML metadata should use this callback url (the one with
the client_name parameter).

Best regards,
Jérôme


2015-07-29 15:47 GMT+02:00 Nuno Alexandre <notifications@github.com>:

**Comment by greven** [comment]
Hi,

That's the thing it does. The IdP has /callback?client_name=Saml2Client (Still using 1.7.0 in dev I have other branch where I'm trying to integrate 1.7.1).

Although the IdP has the correct callback registered in the SP Metadata, once it redirects back to my application I have that error.

The IdP after asserting the request redirects back to my application at the correct endpoint **/callback?client_name=Saml2Client** but that error is thrown.

I tried to replicate it in the pac4j-sparkdemo but the SAML request errored at the IdP level (probably the metadata not registered or something).

**Comment by leleuj** [comment]
To test the SAML integration, you can re-use the SAML support defined in the https://github.com/pac4j/j2e-pac4j-demo. This one works.

**Comment by greven** [comment]
Hi again Jérôme,

Just submitted a pull request with the integration suggested by you. I get an error with the WebContext casting from spark to j2e.

**Comment by leleuj** [comment]
Hi,

Hallelujah! I know what's wrong and it's a bit tricky. To make things
"easier", objects are lazzy initialized: the Clients but each Client as
well, this happens when you try to use them.

In your case, you explicitly need to initialize the samlClient:

public static void main(String[] args) {
   setPort(8080);
   final Clients clients = new ClientsBuilder().build(null);

   SAML2Client samlclient = clients.findClient(SAML2Client.class);
   samlclient.init();
   String spMetadata =
samlclient.getServiceProviderMetadataResolver().getMetadata();
   logger.info(spMetadata);

   get("/", (rq, rs) -> index(rq, rs, clients), templateEngine);
   final Route callback = new CallbackRoute(clients);
   get("/callback", callback);
   post("/callback", callback);
   before("/facebook", new RequiresAuthenticationFilter(clients,
"FacebookClient"));


Thanks for your patience on this.
Best regards,
Jérôme


2015-08-03 14:48 GMT+02:00 Nuno Alexandre <notifications@github.com>:

**Comment by greven** [comment]
Hi Jérôme, that is an issue yes that you can solve with initialize. But the endpoint error is not related to this. Anyway I was trying to integrate the same way I did with the pac4j-spark demo, 1.7.1 with spark-pac4j into my app, but I'm getting an error parsing the IdP metadata on the unmarshaller of the IdP metadata which it didn't occur on 1.7.0. I'm debugging to make sure what it is.
Once I have results I'll report back to see if I can find the source of this and the routing problem.

Thanks for your aid.

**Comment by greven** [comment]
I've found the problem.
I can't test it with the spark demo project since I can't reach the part where spark received the SAML response from the IdP but I can tell you what is happening on my app.

I get an error on BaseSAMLMessageDecoder.class
Line 91.

    if(!matched) {
                this.log.error("SAML message intended destination endpoint \'{}\' did not match the recipient endpoint \'{}\'", messageDestination, receiverEndpoint);
                throw new SecurityException("SAML message intended destination endpoint did not match recipient endpoint");
            } else {
                this.log.debug("SAML message intended destination endpoint matched recipient endpoint");
            }

My SP metadata has this as location (callback url): Location="https://localhost:8888/callback?client_name=Saml2Client"

It is correct. Problem is at that point in the code when comparing to see if the **messageDestination** matches the **receiverEndpoint** it returns false and the Exception "SAML message intended destination endpoint did not match recipient endpoint" is thrown.

What is happening is at this point in the code is not taking into account the Cient Query, in this case the SAML client query: ?client_name=Saml2Client.

So it is comparing https://localhost:8888/callback?client_name=Saml2Client with https://localhost:8888/callback, when it should add the client query to the **receiverEndPoint**

If I set **matched** to true when debugging it will work perfectly.

My authentication code is defined like this:

            final Clients clients = ClientsBuilder.build();

        final Route callback = new CallbackRoute(clients);
        get("/callback", callback);
        post("/callback", callback);

        // Require authentication before proceeding with the rest of the process
        before("/", new RequiresAuthenticationFilter(clients, "Saml2Client"));
        get("/", (req, res) -> postAuthenticationRequest(req, res, clients)); // After authentication

And the client like this:

        public static Clients build() {

        final Saml2Client sapidSaml2Client = new Saml2Client();
        sapidSaml2Client.setKeystorePath(PropertiesHandler.getWebServerProperty("keystore.path"));
        sapidSaml2Client.setKeystorePassword(PropertiesHandler.getWebServerProperty("keystore.pass"));
        sapidSaml2Client.setPrivateKeyPassword(PropertiesHandler.getWebServerProperty("keystore.privatekey.pass"));
        sapidSaml2Client.setIdpMetadataPath("resource:sapid_idp.xml");
        sapidSaml2Client.setSpEntityId(PropertiesHandler.getWebServerProperty("webserver.host"));
        sapidSaml2Client.setMaximumAuthenticationLifetime(3600);

        return new Clients("https://localhost/callback, sapidSaml2Client);

    }

**Comment by bmts** [comment]
Hi,
Any progress on this issue?
Am also having the problem...
Thx

**Comment by greven** [comment]
Hi again Jérôme,

I've continue to investigate to check what might be causing this problem.

Before the code above, before **matched**  we have this line on BasedSAMLMessageDecoder.class:

    String receiverEndpoint = DatatypeHelper.safeTrimOrNullString(this.getActualReceiverEndpointURI(messageContext));

What returns from this line is strippin the query of the client.
So instead of we having "https://localhost:8888/callback?client_name=Saml2Client" we get "https://localhost:8888/callback" and the comparison that I've showed above fails.

Going deeper we get to SparkWebContext.java.
This is where we get:

    	@Override
	public String getFullRequestURL() {
		return request.url();
	}

I guess this function should be something like this instead:


    	@Override
	public String getFullRequestURL() {
		return this.request.url() + "?client__name=" + this.request.queryMap("client_name").value();
	}


Can I do a PR in order for you to check if it works?

Thanks.

**Comment by leleuj** [comment]
Well done! Indeed, the `getFullRequestUrl` method is meant to return the query string as well. You can inspire yourself from the J2E implementation:

```
public String getFullRequestURL() {
  StringBuffer requestURL = request.getRequestURL();
  String queryString = request.getQueryString();
  if (queryString == null) {
    return requestURL.toString();
  } else {
    return requestURL.append('?').append(queryString).toString();
  }
}
```

A PR will be appreciated.

**Comment by leleuj** [comment]
Indeed, there is a cast problem (between `J2EContext` and `SparkWebContext`) using pac4j 1.7.1, I'll fix that in pac4j v1.8.
